oauth2-authorization-server/src/test/java/org/springframework/security/oauth2/server/authorization/authentication/OAuth2RefreshTokenAuthenticationProviderTest.java

```diff
@@ -0,0 +1,266 @@
+
+import static org.assertj.core.api.AssertionsForClassTypes.assertThatThrownBy;
+import static org.assertj.core.api.AssertionsForInterfaceTypes.assertThat;
+import static org.mockito.ArgumentMatchers.any;
+import static org.mockito.Mockito.mock;
+import static org.mockito.Mockito.verify;
+import static org.mockito.Mockito.when;
+
+/**
+ * @author Alexey Nesterov
+ * @since 0.0.3
+ */
+public class OAuth2RefreshTokenAuthenticationProviderTest {
+
+	private final RegisteredClient registeredClient = TestRegisteredClients.registeredClient2().build();
+	private final String NEW_ACCESS_TOKEN_VALUE = UUID.randomUUID().toString();
+	private final String REFRESH_TOKEN_VALUE = UUID.randomUUID().toString();
+
+	private final OAuth2ClientAuthenticationToken clientPrincipal = new OAuth2ClientAuthenticationToken(this.registeredClient);
+	private final OAuth2RefreshTokenAuthenticationToken authentication = new OAuth2RefreshTokenAuthenticationToken(clientPrincipal, REFRESH_TOKEN_VALUE);
+
+	private final OAuth2AccessToken existingAccessToken = new OAuth2AccessToken(
+			OAuth2AccessToken.TokenType.BEARER,
+			"old-test-access-token",
+			Instant.now(),
+			Instant.now().plusSeconds(10),
+			this.registeredClient.getScopes());
+
+	private final OAuth2Authorization authorization = OAuth2Authorization
+															.withRegisteredClient(this.registeredClient)
+															.principalName(this.registeredClient.getClientId())
+															.tokens(OAuth2Tokens.builder()
+																			.accessToken(existingAccessToken)
+																			.refreshToken(new OAuth2RefreshToken(REFRESH_TOKEN_VALUE, Instant.now()))
+																			.build())
+															.build();
+
+	private OAuth2AuthorizationService authorizationService;
+	private JwtEncoder jwtEncoder;
+	private OAuth2RefreshTokenAuthenticationProvider provider;
+
+	@Before
+	public void setUp() {
+		this.authorizationService = mock(OAuth2AuthorizationService.class);
+		this.jwtEncoder = mock(JwtEncoder.class);
+		this.provider = new OAuth2RefreshTokenAuthenticationProvider(this.authorizationService, this.jwtEncoder);
+
+		Jwt jwt = Jwt.withTokenValue(NEW_ACCESS_TOKEN_VALUE)
+						.issuedAt(Instant.now())
+						.header(JoseHeaderNames.ALG, SignatureAlgorithm.RS256.getName())
+						.build();
+
+		when(this.jwtEncoder.encode(any(), any())).thenReturn(jwt);
+	}
+
+	@Test
+	public void supportsWhenSupportedAuthenticationThenTrue() {
+		assertThat(this.provider.supports(OAuth2RefreshTokenAuthenticationToken.class)).isTrue();
+	}
+
+	@Test
+	public void supportsWhenSupportedAuthenticationThenFalse() {
+		assertThat(this.provider.supports(OAuth2ClientCredentialsAuthenticationToken.class)).isFalse();
+	}
+
+	@Test
+	public void authenticateWhenRefreshTokenExistsThenReturnAuthentication() {
+		when(this.authorizationService.findByToken(REFRESH_TOKEN_VALUE, TokenType.REFRESH_TOKEN))
+				.thenReturn(authorization);
+
+		OAuth2AccessTokenAuthenticationToken accessTokenAuthentication =
```

Comment:
Let's create `OAuth2RefreshTokenAuthenticationToken` for each test instead of using the member